src/zino/planned_maintenance.py

```diff
@@ -110,44 +119,95 @@ def periodic(self):
 
         now = datetime.datetime.now()
 
-        started_maintenances = self.get_started_planned_maintenances(now=now)
-        for started_pm in started_maintenances:
-            events = self._get_or_create_maintenance_events_for_maintenance_start(maintenance=started_pm)
-            for event in events:
-                self._ignore_event(event)
-                started_pm.event_ids.append(event.id)
+        # Initiate PM once it becomes active
+        for started_pm in self.get_started_planned_maintenances(now=now):
+            self._start(started_pm)
 
-        active_maintenances = self.get_active_planned_maintenances(now=now)
-        # Get all events (maybe already filter here for ignored and closed events?)
+        # Make sure all events that matches a PM is ignored
         for event in state.events:
-            for active_pm in active_maintenances:
-                if active_pm.matches(event):
-                    self._ignore_event(event)
-                    active_pm.event_ids.append(event.id)
-
-        ended_maintenances = self.get_ended_planned_maintenances(now=now)
-        for ended_pm in ended_maintenances:
-            for event in ended_pm.pm_events:
-                if event.state is not EventState.OPEN:
-                    self._unignore_event(event)
-
-    def _ignore_event(self, event: Event):
+            self._check_event(event)
+
+        # End a PM and set events matching the PM to open
+        for ended_pm in self.get_ended_planned_maintenances(now=now):
+            self._end(ended_pm)
+
+        self.last_run = now
+
+    def _start(self, pm: PlannedMaintenance):
         from zino.state import state
 
-        event.state = EventState.IGNORED
-        state.events.commit(event)
+        events = self._get_or_create_events(pm)
+        for event in events:
+            event.state = EventState.IGNORED
+            state.events.commit(event)
+            pm.event_ids.append(event.id)
```

Comment:
thats my understanding of how it works, since that will catch any alarm that occurs during the PM timeframe even if it didnt exist before. tcl isnt the easiest to read tho..

Reply:
No, I think this is how it is done today. So we should do it like this as well, but put the eventual improvement on the list for further development